bacon/model/StorageCouchBaseDB.swift

```diff
@@ -0,0 +1,136 @@
+//
+//  StorageMongoDB.swift
+//  bacon
+//
+//  Underlying class of StorageManager to provide all saving and loading functionalities.
+//  Uses CouchBaseLiteSwfit to provide a full-featured embedded NoSQL database that runs locally.
+//
+//  Created by Travis Ching Jia Yea on 19/3/19.
+//  Copyright © 2019 nus.CS3217. All rights reserved.
+//
+
+import Foundation
+import CouchbaseLiteSwift
+
+private enum DatabaseCollections: String {
+    case transactions
+}
+
+class StorageCouchBaseDB {
+    private var transactionDatabase: Database
+
+    init() throws {
+        // Initialize database
+        transactionDatabase = try StorageCouchBaseDB.openOrCreateEmbeddedDatabase(name: .transactions)
+    }
+
+    private static func openOrCreateEmbeddedDatabase(name: DatabaseCollections) throws -> Database {
+        do {
+            let options = DatabaseConfiguration()
+            // Get the path to the Database
+            let urls = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
+            guard let documentDirectory = urls.first else {
+                throw InitializationError(message: "unable to access document directory!")
+            }
+            let databaseFolderUrl = documentDirectory.appendingPathComponent(name.rawValue, isDirectory: true)
+            let databaseFolderPath = try getOrCreateFolderPath(for: databaseFolderUrl)
+            // Set the folder path for the CouchbaseLite Database
+            options.directory = databaseFolderPath
+            // Create a new database or get handle to existing database at specified path
+            return try Database(name: name.rawValue, config: options)
+        } catch let error {
+            if error is InitializationError {
+                throw error
+            } else {
+                throw InitializationError(message: "unable to access embedded \(name.rawValue) mobile database.")
+            }
+        }
+    }
+
+    private static func getOrCreateFolderPath(for url: URL) throws -> String {
+        let folderPath = url.path
+        let fileManager = FileManager.default
+        if !fileManager.fileExists(atPath: folderPath) {
+            // Create folder if non-existent
+            do {
+                try fileManager.createDirectory(atPath: folderPath,
+                                                withIntermediateDirectories: true,
+                                                attributes: nil)
+            } catch {
+                throw InitializationError(message:
+                    "Database directory path is non-existent and encountered error creating directory: \(folderPath)")
+            }
+        }
+        return folderPath
+    }
+
+    private func createMutableDocument(from transaction: Transaction) throws -> MutableDocument {
```

Comment:
Is there a reason for the above 2 methods to be static but not this?

Reply:
The static methods are used to create the database instance, but this createMutableDocument is only called when saving.